blocks/graph.py

```diff
@@ -161,8 +162,45 @@ def replace(self, replacements):
             substitutes.
 
         """
-        return ComputationGraph(theano.clone(self.outputs,
-                                             replace=replacements))
+        # Due to theano specifics we have to make one replacement in time
+        replacements = OrderedDict(replacements)
+
+        # `replacements` with previous replacements applied. We have to track
+        # variables in the new graph corresponding to original replacements.
+        replacements_cur = replacements
+
+        # Maps original `replacement` keys to
+        # new, replaced earlier, keys in `replacements_cur`.
+        key_to_key_cur = OrderedDict(zip(replacements.keys(),
+                                         replacements_cur.keys()))
+        apply_nodes_sorted = theano.gof.graph.io_toposort(
+            self.inputs, self.outputs)
+        outputs_cur = self.outputs
+        # Replace step-by-step in topological order
+        for node in apply_nodes_sorted:
+            for input_ in node.inputs:
+                if input_ not in replacements:
+                    continue
+                # Find a source/target to replace during this step.
+                replace_what = key_to_key_cur[input_]
+                replace_by = replacements_cur[key_to_key_cur[input_]]
+
+                # We also want to make changes in future replacements
+                outputs_new = theano.clone(
+                    outputs_cur +
+                    list(replacements_cur.keys()) +
+                    list(replacements_cur.values()),
+                    replace={replace_what: replace_by})
+                # Reconstruct outputs, keys, and values
+                outputs_cur = outputs_new[:len(outputs_cur)]
+                keys = outputs_new[len(outputs_cur):
+                                   len(outputs_cur) + len(replacements)]
+                vals = outputs_new[len(outputs_cur) + len(replacements):]
+                # Update mappings
+                replacements_cur = OrderedDict(zip(keys, vals))
+                key_to_key_cur = OrderedDict(zip(keys,
```

Comment:
First of all, thanks for your comments and renamings, it is much easier to read this code now.

It seems to me that `key_to_key_cur` will always have identical keys and values from these two lines.